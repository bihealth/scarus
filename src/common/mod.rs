//! Commonly used code.

use clap::Parser;
use clap_verbosity_flag::{InfoLevel, Verbosity};

/// Commonly used command line arguments.
#[derive(Parser, Debug)]
pub struct Args {
    /// Verbosity of the program
    #[clap(flatten)]
    pub verbose: Verbosity<InfoLevel>,
}
