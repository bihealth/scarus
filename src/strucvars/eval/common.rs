//! Common functions and data strucdtures for evaluation.

/// Scores of the results in individual categories.
pub trait SuggestedScore {
    /// Suggested score for the category.
    fn suggested_score(&self) -> f32;
}

/// Score range for a seciton.
pub trait ScoreRange {
    /// Minimal score for the category.
    fn min_score(&self) -> f32;

    /// Maximum score for the category.
    fn max_score(&self) -> f32;
}
