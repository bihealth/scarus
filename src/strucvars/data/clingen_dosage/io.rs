//! I/O implementation for `ClinGen` dosage sensitivity.

use std::{
    io::{BufRead, BufReader},
    path::Path,
};

use crate::common::Assembly;

/// Helper trait to adjust the region to the chromosome.
pub trait AdjustForAssembly {
    /// Adjust the region to the chromosome.
    fn for_assembly(self, assembly: Assembly) -> Self;
}

/// `ClinGen` gene dosage sensitivity TSV entry.
#[derive(Debug, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct Gene {
    /// Gene symbol.
    #[serde(alias = "#Gene Symbol")]
    pub gene_symbol: String,
    /// NCBI gene ID.
    #[serde(alias = "Gene ID")]
    pub ncbi_gene_id: String,
    /// Genomic location.
    #[serde(alias = "Genomic Location")]
    pub genomic_location: String,
    /// Haploinsufficiency score.
    #[serde(alias = "Haploinsufficiency Score", deserialize_with = "parse_score")]
    pub haploinsufficiency_score: Option<u32>,
    /// Triplosensitivity score.
    #[serde(alias = "Triplosensitivity Score", deserialize_with = "parse_score")]
    pub triplosensitivity_score: Option<u32>,
    /// Haploinsufficiency Disease ID.
    #[serde(alias = "Haploinsufficiency Disease ID")]
    pub haploinsufficiency_disease_id: Option<String>,
    /// Haploinsufficiency Disease ID.
    #[serde(alias = "Triplosensitivity Disease ID")]
    pub triplosensitivity_disease_id: Option<String>,
}

impl TryInto<super::Gene> for Gene {
    type Error = anyhow::Error;

    fn try_into(self) -> Result<super::Gene, Self::Error> {
        Ok(super::Gene {
            gene_symbol: self.gene_symbol,
            ncbi_gene_id: self.ncbi_gene_id,
            genomic_location: self.genomic_location,
            haploinsufficiency_score: super::Score::try_from(self.haploinsufficiency_score)
                .map_err(|e| anyhow::anyhow!("invalid haploinsufficiency score: {}", e))?,
            triplosensitivity_score: super::Score::try_from(self.triplosensitivity_score)
                .map_err(|e| anyhow::anyhow!("invalid haploinsufficiency score: {}", e))?,
            haploinsufficiency_disease_id: self.haploinsufficiency_disease_id,
            triplosensitivity_disease_id: self.triplosensitivity_disease_id,
        })
    }
}

impl AdjustForAssembly for Gene {
    fn for_assembly(self, assembly: Assembly) -> Self {
        Self {
            genomic_location: if assembly == Assembly::Grch37 {
                self.genomic_location.replace("chr", "")
            } else {
                self.genomic_location
            },
            ..self
        }
    }
}

/// `ClinGen` region dosage sensitivity entry.
#[derive(Debug, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct Region {
    /// ISCA ID
    #[serde(alias = "#ISCA ID")]
    pub isca_id: String,
    /// ISCA Region Name
    #[serde(alias = "ISCA Region Name")]
    pub isca_region_name: String,
    /// Genomic location.
    #[serde(alias = "Genomic Location")]
    pub genomic_location: String,
    /// Haploinsufficiency score.
    #[serde(alias = "Haploinsufficiency Score", deserialize_with = "parse_score")]
    pub haploinsufficiency_score: Option<u32>,
    /// Triplosensitivity score.
    #[serde(alias = "Triplosensitivity Score", deserialize_with = "parse_score")]
    pub triplosensitivity_score: Option<u32>,
    /// Haploinsufficiency Disease ID.
    #[serde(alias = "Haploinsufficiency Disease ID")]
    pub haploinsufficiency_disease_id: Option<String>,
    /// Haploinsufficiency Disease ID.
    #[serde(alias = "Triplosensitivity Disease ID")]
    pub triplosensitivity_disease_id: Option<String>,
}

impl TryInto<super::Region> for Region {
    type Error = anyhow::Error;

    fn try_into(self) -> Result<super::Region, Self::Error> {
        Ok(super::Region {
            isca_id: self.isca_id,
            isca_region_name: self.isca_region_name,
            genomic_location: self.genomic_location,
            haploinsufficiency_score: super::Score::try_from(self.haploinsufficiency_score)
                .map_err(|e| anyhow::anyhow!("invalid haploinsufficiency score: {}", e))?,
            triplosensitivity_score: super::Score::try_from(self.triplosensitivity_score)
                .map_err(|e| anyhow::anyhow!("invalid haploinsufficiency score: {}", e))?,
            haploinsufficiency_disease_id: self.haploinsufficiency_disease_id,
            triplosensitivity_disease_id: self.triplosensitivity_disease_id,
        })
    }
}

impl AdjustForAssembly for Region {
    fn for_assembly(self, assembly: Assembly) -> Self {
        Self {
            genomic_location: if assembly == Assembly::Grch37 {
                self.genomic_location.replace("chr", "")
            } else {
                self.genomic_location
            },
            ..self
        }
    }
}

/// Helper for parsing the scores which may have interesting values.
fn parse_score<'de, D>(d: D) -> Result<Option<u32>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    let tmp: String = serde::Deserialize::deserialize(d)?;
    if tmp.is_empty() || tmp == "Not yet evaluated" || tmp == "-1" {
        Ok(None)
    } else {
        Ok(Some(tmp.parse().map_err(serde::de::Error::custom)?))
    }
}

/// Load `ClinGen` regions file.
///
/// # Arguments
///
/// * `path` - Path to `ClinGen` regions file.
/// * `assembly` - Assembly of the regions file, used to determine whether or not to
///   strip the `chr` prefix from the chromosome name.
///
/// # Returns
///
/// `ClinGen` regions.
///
/// # Errors
///
/// If anything goes wrong, it returns a generic `anyhow::Error`.
pub fn load_file<T, P>(path: P, assembly: Assembly) -> Result<Vec<T>, anyhow::Error>
where
    T: serde::de::DeserializeOwned + std::fmt::Debug + AdjustForAssembly,
    P: AsRef<Path>,
{
    // Construct reader and skip initial 5 lines.
    let mut reader = std::fs::File::open(path)
        .map_err(|e| anyhow::anyhow!("problem opening file: {}", e))
        .map(BufReader::new)?;

    {
        let mut buf = String::new();
        for _ in 0..5 {
            reader.read_line(&mut buf)?;
            buf.clear();
        }
    }

    let mut csv_reader = csv::ReaderBuilder::new()
        .delimiter(b'\t')
        .has_headers(true)
        .flexible(true)
        .from_reader(reader);
    let mut result = Vec::new();
    for record in csv_reader.deserialize() {
        let record: T = record.map_err(|e| anyhow::anyhow!("problem parsing record: {}", e))?;
        result.push(record.for_assembly(assembly));
    }

    Ok(result)
}

#[cfg(test)]
mod test {
    use crate::common::Assembly;

    #[test]
    fn load_clingen_genes_file_grch37() -> Result<(), anyhow::Error> {
        let genes = super::load_file::<super::Gene, _>(
            "tests/data/strucvars/ClinGen_gene_curation_list_GRCh37.tsv",
            Assembly::Grch37,
        )?;

        assert_eq!(genes.len(), 1518);
        insta::assert_yaml_snapshot!(&genes[0..5]);

        Ok(())
    }

    #[test]
    fn load_clingen_regions_file_grch37() -> Result<(), anyhow::Error> {
        let regions = super::load_file::<super::Region, _>(
            "tests/data/strucvars/ClinGen_region_curation_list_GRCh37.tsv",
            Assembly::Grch37,
        )?;

        assert_eq!(regions.len(), 513);
        insta::assert_yaml_snapshot!(&regions[0..5]);

        Ok(())
    }

    #[test]
    fn load_clingen_regions_file_grch38() -> Result<(), anyhow::Error> {
        let regions = super::load_file::<super::Region, _>(
            "tests/data/strucvars/ClinGen_region_curation_list_GRCh38.tsv",
            Assembly::Grch38,
        )?;

        assert_eq!(regions.len(), 514);
        insta::assert_yaml_snapshot!(&regions[0..5]);

        Ok(())
    }

    #[test]
    fn gene_conversion() -> Result<(), anyhow::Error> {
        let io_gene = super::Gene {
            gene_symbol: "AASS".to_string(),
            ncbi_gene_id: "53947".to_string(),
            genomic_location: "chr7:121713598-121784344".to_string(),
            haploinsufficiency_score: Some(30),
            triplosensitivity_score: None,
            haploinsufficiency_disease_id: Some("MONDO:0009388".to_string()),
            triplosensitivity_disease_id: None,
        };

        let gene: super::super::Gene = io_gene.try_into()?;

        assert_eq!(
            gene,
            super::super::Gene {
                gene_symbol: "AASS".to_string(),
                ncbi_gene_id: "53947".to_string(),
                genomic_location: "chr7:121713598-121784344".to_string(),
                haploinsufficiency_score: super::super::Score::GeneAssociatedWithRecessivePhenotype,
                triplosensitivity_score: super::super::Score::NoEvidenceAvailable,
                haploinsufficiency_disease_id: Some("MONDO:0009388".to_string()),
                triplosensitivity_disease_id: None,
            }
        );

        Ok(())
    }

    #[test]
    fn region_conversion() -> Result<(), anyhow::Error> {
        let io_region = super::Region {
            isca_id: "ISCA-46748".to_string(),
            isca_region_name: "Xq25 region (includes STAG2)".to_string(),
            genomic_location: "chrX:123034319-123236519".to_string(),
            haploinsufficiency_score: Some(3),
            triplosensitivity_score: Some(3),
            haploinsufficiency_disease_id: Some("MONDO:0100038".to_string()),
            triplosensitivity_disease_id: None,
        };

        let region: super::super::Region = io_region.try_into()?;

        assert_eq!(
            region,
            super::super::Region {
                isca_id: "ISCA-46748".to_string(),
                isca_region_name: "Xq25 region (includes STAG2)".to_string(),
                genomic_location: "chrX:123034319-123236519".to_string(),
                haploinsufficiency_score: super::super::Score::SufficientEvidence,
                triplosensitivity_score: super::super::Score::SufficientEvidence,
                haploinsufficiency_disease_id: Some("MONDO:0100038".to_string()),
                triplosensitivity_disease_id: None,
            }
        );

        Ok(())
    }
}
