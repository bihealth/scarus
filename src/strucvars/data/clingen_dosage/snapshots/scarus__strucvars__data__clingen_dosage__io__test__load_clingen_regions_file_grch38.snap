---
source: src/strucvars/data/clingen_dosage/io.rs
expression: "&regions[0..5]"
---
- isca_id: ISCA-46748
  isca_region_name: 19p13.3 Critical Region 1 (CR1)
  genomic_location: tbd
  haploinsufficiency_score: 0
  triplosensitivity_score: 1
  haploinsufficiency_disease_id: ~
  triplosensitivity_disease_id: "MONDO:0100038"
- isca_id: ISCA-46747
  isca_region_name: "22q11.2 recurrent region (distal type III, E/F-H) (includes SMARCB1)"
  genomic_location: "chr22:22776925-24598466"
  haploinsufficiency_score: 3
  triplosensitivity_score: 1
  haploinsufficiency_disease_id: ~
  triplosensitivity_disease_id: ~
- isca_id: ISCA-46743
  isca_region_name: Xq25 region (includes STAG2)
  genomic_location: "chrX:123900469-124102669"
  haploinsufficiency_score: 3
  triplosensitivity_score: 3
  haploinsufficiency_disease_id: ~
  triplosensitivity_disease_id: "MONDO:0100038"
- isca_id: ISCA-46742
  isca_region_name: 7p22.1 region (includes ACTB)
  genomic_location: "chr7:5497217-5760091"
  haploinsufficiency_score: 3
  triplosensitivity_score: 1
  haploinsufficiency_disease_id: "MONDO:0000508"
  triplosensitivity_disease_id: "MONDO:0100038"
- isca_id: ISCA-46741
  isca_region_name: 9p24.3 Region (includes DMRT1)
  genomic_location: "chr9:883300-918342"
  haploinsufficiency_score: 2
  triplosensitivity_score: ~
  haploinsufficiency_disease_id: "MONDO:0002145"
  triplosensitivity_disease_id: ~

